### PR TITLE
Updated Report Generation Model

### LINKED ISSUE #64: Keep inference outside reporting
@michaelkiper Opening a thread to discuss the inference/report structure; my main argument is that each model requires a different way to do inference, so we can't really abstract it inside the report wrapper. SpecTf needs minibatching for inference with large enough datasets, for example, and there's a laundry list of other considerations (CPU or GPU? tensor or ndarray? ndims?)

A very simple way to do this would be to simply make `self.y_hat` a variable you need to set before exiting the context, as I have done in the `spectf_lee` branch:

https://github.com/emit-sds/cover-class/blob/faada6ea717272015f4b595ac4bfe90c817e2d79/src/cover_class/reporting/report_config.py#L49

Then `make_report()` runs with the provided y_hat instead of trying to do its own inference:

https://github.com/emit-sds/cover-class/blob/faada6ea717272015f4b595ac4bfe90c817e2d79/src/cover_class/reporting/report_config.py#L100-L108

This encourages users to cast their predictions to the correct device/ndarray/datatype/ndims before it ever reaches the report context, which is good, I think.

### PR DESCRIPTION
**Pull request by michaelkiper**
## Overview
This PR now changes the report generation method to now be an invokable function instead of an automatically occurring function at the end of a context window. The reason being is that users want to be able to control the way testing inference is done at a more granular level.

Please see the updated `README.md` and `models/training_example.py` for a user's guide.

## Related Tickets
resolves https://github.com/emit-sds/cover-class/issues/64

## Testing
Ran regression testing with `make test`

Then did a smoke test with the example script: `python3 models/training_example.py --outdir . --config config/dataloaderyml --lr 0.0001 --batch-size 1024 --max-training-steps 100 --log-interval 20 --simulated-test-set-size 1000 --seed 42` which generated the expected output.

Thresholds are now reported as part of the class items:
Screenshot 2026-01-05 at 5 16 18 PM

### CODE REVIEW

File: src/cover_class/reporting/report_config.py [hunk 1]
```diff
@@ -43,10 +43,8 @@ class Report:
     outdir: str
     config: str|Dict
     model_config: ModelConfig
-    X_test: Union[Tensor, NDArray]
     Y_test: Union[Tensor, NDArray]
 
-    classification_threshold: Optional[List[float]] = None
```
Comment (michaelkiper) [inline src/cover_class/reporting/report_config.py:49]: This should be removed to be in `make_report` now

File: src/cover_class/reporting/report_config.py [hunk 2]
```diff
@@ -88,33 +79,27 @@ def __post_init__(self):
...
-        f1_scores          = f_beta_scores(y_hat, self.Y_test, class_names)
+        
+        assert y_hat.shape[-1] == len(class_names), f"Got {y_hat.shape[-1]} classes in y_hat, but {len(class_names)} classes from the config"
+        assert len(class_thresholds) == len(class_names), f"Got {len(class_thresholds)} class thresholds, but {len(class_names)} classes from the config"
+
+        cm_plot            = confusion_matrix(y_hat_binary, self.Y_test, class_names)
+        mcc_plot           = missed_class_confusion(y_hat_binary, self.Y_test, class_names)
+        rates              = tpr_fpr(y_hat_binary, self.Y_test, class_names)
+        f1_scores          = f_beta_scores(y_hat_binary, self.Y_test, class_names)
         roc_plot, roc_dict = roc_auc(y_hat, self.Y_test, class_names)
```
Comment (michaelkiper) [inline src/cover_class/reporting/report_config.py:100]: There was a bug previously where the ROC was being calculated based on the binary values, not on the continuous probabilities.
